File: Step2_Data_Preprocessing/step2_feature_extraction_synthetic2D_Train.m
%% Synthetic Data Subvolume Extraction and Image Generation
% This script processes synthetic atomistic datasets, divides them into subvolumes, 
% and generates 2D projections (XY, XZ, YZ planes) for CNN training.

%% ============================= USER INPUTS =============================
baseFolder = '/home/tangy/cluster/Synthetic/';  % <-- Change this to your base folder
outputFolder = 'Train';   % <-- Folder to save generated images
numSyntheticFiles = 630;  % <-- Total number of synthetic datasets
cubeSize = 20;            % <-- Subvolume size in nm
gridSize = 100;           % <-- Grid size for projection images
minPointsThreshold = 300; % <-- Minimum number of points in subvolume to process
minClusterSize = 100;     % <-- Minimum cluster size to consider
resizeImageSize = [100, 100]; % <-- Final image size (pixels)
%% ======================================================================

iter = 1;  % Image counter
parameter = {}; 
groundtruth = [];

for syn = 1:numSyntheticFiles
    % Construct paths for input synthetic dataset
    folderName = ['GeneratedDatasetResults_Synthetic', num2str(syn)];
    fileName = ['Synthetic', num2str(syn), '_CuAtomsInClusterandSolidSolution.txt'];
    path = fullfile(baseFolder, folderName, fileName);

    % Read the dataset
    Data = readmatrix(path);

    % Compute number of cubes in x, y, z directions
    num_cubes = floor((max(Data) - min(Data)) / cubeSize);

    % Divide dataset into subvolumes
    data = cell(num_cubes);  % Preallocate
    for i = 1:num_cubes(1)
        for j = 1:num_cubes(2)
            for k = 1:num_cubes(3)
                x = min(Data(:, 1)) + cubeSize * (i - 1);
                y = min(Data(:, 2)) + cubeSize * (j - 1);
                z = min(Data(:, 3)) + cubeSize * (k - 1);

                log1 = (x <= Data(:, 1) & Data(:, 1) <= (x + cubeSize));
                log2 = (y <= Data(:, 2) & Data(:, 2) <= (y + cubeSize));
                log3 = (z <= Data(:, 3) & Data(:, 3) <= (z + cubeSize));
                data{i, j, k} = Data(log1 & log2 & log3, :);
            end
        end
    end

    % Process each subvolume
    for i = 1:size(data, 1)
        for j = 1:size(data, 2)
            for k = 1:size(data, 3)
                gt = data{i, j, k};

                % Remove invalid points (label = -1)
                gt(gt(:, 5) == -1, :) = [];

                if ~isempty(gt) && size(gt, 1) > minPointsThreshold
                    % Plot for visual check (optional)
                    % figure; scatter3(gt(:, 1), gt(:, 2), gt(:, 3), 5, '.')

                    % Count clusters
                    [uniqueVals, ~, idx] = unique(gt(:, 5));
                    counts = accumarray(idx, 1);
                    t = [uniqueVals, counts];
                    t_filtered = t(t(:, 2) >= minClusterSize, :);

                    cluster_num = size(t_filtered, 1);

                    % Scale coordinates for grid representation
                    minCoords = min(gt(:, 1:3));
                    maxCoords = max(gt(:, 1:3));
                    scaledCoords = (gt(:, 1:3) - minCoords) ./ (maxCoords - minCoords) * (gridSize - 1);

                    % Initialize 2D projections
                    Count_xy = zeros(gridSize, gridSize);
                    Count_xz = zeros(gridSize, gridSize);
                    Count_yz = zeros(gridSize, gridSize);

                    % Populate projections
                    for pt = 1:size(gt, 1)
                        ix = floor(scaledCoords(pt, 1)) + 1;
                        iy = floor(scaledCoords(pt, 2)) + 1;
                        iz = floor(scaledCoords(pt, 3)) + 1;
                        Count_xy(ix, iy) = Count_xy(ix, iy) + 1;
                        Count_xz(ix, iz) = Count_xz(ix, iz) + 1;
                        Count_yz(iy, iz) = Count_yz(iy, iz) + 1;
                    end

                    % Save projections as images
                    filename = fullfile(outputFolder, [num2str(iter), '.jpg']);
                    imwrite(count_rgb(Count_xy, resizeImageSize), filename);
                    parameter{iter} = [cluster_num];
                    groundtruth(iter, :) = [syn, i, j, k, cluster_num];
                    iter = iter + 1;

                    filename = fullfile(outputFolder, [num2str(iter), '.jpg']);
                    imwrite(count_rgb(Count_xz, resizeImageSize), filename);
                    parameter{iter} = [cluster_num];
                    groundtruth(iter, :) = [syn, i, j, k, cluster_num];
                    iter = iter + 1;

                    filename = fullfile(outputFolder, [num2str(iter), '.jpg']);
                    imwrite(count_rgb(Count_yz, resizeImageSize), filename);
                    parameter{iter} = [cluster_num];
                    groundtruth(iter, :) = [syn, i, j, k, cluster_num];
                    iter = iter + 1;
                end
            end
        end
    end
end

%% ============================= HELPER FUNCTION =============================
function img_resized = count_rgb(img, resizeImageSize)
    % Convert matrix to RGB image and resize
    figure; imagesc(img);
    clim([0, max(img(:))]);
    colormap_used = colormap;
    frame = getframe(gca);
    im_rgb = frame2im(frame);
    img_resized = imresize(im_rgb, resizeImageSize);
    close;
end

File: Step2_Data_Preprocessing/Step2_feature_extraction_synthetic2D_Train.py
# Synthetic Data Subvolume Extraction and Image Generation (Python)
# ---------------------------------------------------------------

import os
import math
import numpy as np
from PIL import Image
from matplotlib import cm
from matplotlib import colors
from scipy.io import loadmat


# ============================= USER INPUTS =============================
base_folder = "./SyntheticData"   # <-- change to your base folder
output_folder = "Train"                           # <-- folder to save generated images
num_synthetic_files = 1                           # <-- total number of synthetic datasets
cube_size = 20.0                                  # <-- subvolume size in nm
grid_size = 100                                   # <-- grid size for projection images
min_points_threshold = 300                        # <-- minimum number of points in subvolume to process
min_cluster_size = 100                            # <-- minimum cluster size to consider
resize_image_size = (100, 100)                    # <-- final image size (pixels) (width, height)

# ======================================================================
 
os.makedirs(output_folder, exist_ok=True)
 
iter_counter = 1
parameters = []          # to mirror MATLAB "parameter" cell array (stores cluster_num)
groundtruth = []         # rows: [syn, i, j, k, cluster_num]
 
def count_rgb(img: np.ndarray, out_size=(100, 100)) -> Image.Image:
    # """
    # Emulates MATLAB's imagesc + current colormap -> RGB, then resizes.
    # - Uses matplotlib's 'viridis' by default (stable, perceptually uniform).
    # - Scales to [0, max(img)], handling the all-zero case.
    # """
    img = np.asarray(img, dtype=float)
    vmax = float(img.max()) if img.size > 0 else 0.0
    if vmax <= 0.0:
        # All zeros -> produce a black image
        rgba = np.zeros((img.shape[0], img.shape[1], 4), dtype=np.float32)
    else:
        norm = colors.Normalize(vmin=0.0, vmax=vmax, clip=True)
        cmap = cm.get_cmap("viridis")  # pick a fixed, nice colormap
        rgba = cmap(norm(img))         # floats in [0,1], shape (H,W,4)
 
    rgb = (rgba[..., :3] * 255.0).astype(np.uint8)
    pil_img = Image.fromarray(rgb, mode="RGB")
    # MATLAB imresize default is bicubic-like; Pillow.BILINEAR is a reasonable match
    return pil_img.resize(out_size, resample=Image.BILINEAR)


for syn in range(1, num_synthetic_files + 1):
#     folder_name = f"GeneratedDatasetResults_Synthetic{syn}"
#     file_name = f"Synthetic{syn}_CuAtomsInClusterandSolidSolution.txt"
   file_name='Cr_clusters.txt'
   path = os.path.join(base_folder, file_name)


   # Read the dataset (expecting numeric columns; 5th col is label)
   # If your files have headers or different delimiters, adjust loadtxt arguments.
   Data = np.loadtxt(path)
 

   # Work only with spatial columns for cube tiling
   mins = Data[:, :3].min(axis=0)
   maxs = Data[:, :3].max(axis=0)
   spans = maxs - mins

   # Number of cubes along x, y, z
   num_cubes = np.floor(spans / cube_size).astype(int)  # shape (3,)

   # Guard: if any dimension has 0 cubes, nothing to do
   if np.any(num_cubes <= 0):
       continue

   # Divide dataset into subvolumes 
   for i in range(1, num_cubes[0] + 1):
       x0 = mins[0] + cube_size * (i - 1)
       x1 = x0 + cube_size
       in_x = (Data[:, 0] >= x0) & (Data[:, 0] <= x1)

       for j in range(1, num_cubes[1] + 1):
           y0 = mins[1] + cube_size * (j - 1)
           y1 = y0 + cube_size
           in_y = (Data[:, 1] >= y0) & (Data[:, 1] <= y1)

           for k in range(1, num_cubes[2] + 1):
               z0 = mins[2] + cube_size * (k - 1)
               z1 = z0 + cube_size
               in_z = (Data[:, 2] >= z0) & (Data[:, 2] <= z1)

               gt = Data[in_x & in_y & in_z]

               # Remove invalid points (label = -1) — assumes labels are in column 5 (0-based idx 4)
               if gt.size == 0:
                   continue
               gt = gt[gt[:, 2] != -1]

               if gt.shape[0] <= min_points_threshold:
                   continue

               # Count clusters >= min_cluster_size
               labels = gt[:, 3].astype(int)
               # Equivalent to MATLAB: [uniqueVals, ~, idx] + accumarray
               unique_vals, counts = np.unique(labels, return_counts=True)
               t_filtered = counts[counts >= min_cluster_size]
               cluster_num = int(t_filtered.size)

               # Scale coordinates to [0, grid_size - 1]
               sub_min = gt[:, :3].min(axis=0)
               sub_max = gt[:, :3].max(axis=0)
               ranges = sub_max - sub_min
               # avoid division by zero for flat ranges
               ranges[ranges == 0] = 1.0
               scaled = (gt[:, :3] - sub_min) / ranges * (grid_size - 1)

               # Integer indices (MATLAB used floor + 1; here we keep 0-based)
               idxs = np.floor(scaled).astype(int)
               # clamp just in case of numerical edge (e.g., exactly grid_size)
               np.clip(idxs, 0, grid_size - 1, out=idxs)

               # Initialize projections
               Count_xy = np.zeros((grid_size, grid_size), dtype=np.int32)
               Count_xz = np.zeros((grid_size, grid_size), dtype=np.int32)
               Count_yz = np.zeros((grid_size, grid_size), dtype=np.int32)

               # Populate projections (vectorized with np.add.at)
               ix, iy, iz = idxs[:, 0], idxs[:, 1], idxs[:, 2]
               np.add.at(Count_xy, (ix, iy), 1)
               np.add.at(Count_xz, (ix, iz), 1)
               np.add.at(Count_yz, (iy, iz), 1)

               # Save images (three views)
               for mat in (Count_xy, Count_xz, Count_yz):
                   img = count_rgb(mat, out_size=resize_image_size)
                   out_path = os.path.join(output_folder, f"{iter_counter}.jpg")
                   img.save(out_path, format="JPEG", quality=95)

                   parameters.append([cluster_num])
                   groundtruth.append([syn, i, j, k, cluster_num])
                   iter_counter += 1
               

# (Optional) Save metadata like MATLAB variables would
# np.savetxt(os.path.join(output_folder, "parameters.csv"), np.array(parameters, dtype=int), fmt="%d", delimiter=",")
# np.savetxt(os.path.join(output_folder, "groundtruth.csv"), np.array(groundtruth, dtype=int), fmt="%d", delimiter=",")

print(f"Done. Wrote {iter_counter-1} images to: {output_folder}")




